llama_index/prompts/mixin.py

```diff
@@ -0,0 +1,112 @@
+                raise ValueError(f"Prompt key {key} cannot contain ':'.")
+
+        for key in module_dict:
+            if ":" in key:
+                raise ValueError(f"Prompt key {key} cannot contain ':'.")
+
+        # # validate each prompt module to make sure there's no overlapping keys
+        # # (otherwise update will fail)
+        # # start a counter of keys
+        # all_prompt_keys = defaultdict(int)
+        # # add keys from prompt dict
+        # for key in prompts_dict:
+        #     all_prompt_keys[key] += 1
+        # # add keys from each prompt module
+        # for module in module_dict.values():
+        #     for key in module.get_prompts():
+        #         all_prompt_keys[key] += 1
+        # # check for duplicates
+        # for key, count in all_prompt_keys.items():
+        #     if count > 1:
+        #         raise ValueError(f"Duplicate prompt key {key} found.")
+
+    def get_prompts(self) -> Dict[str, BasePromptTemplate]:
+        """Get a prompt."""
+        prompts_dict = self._get_prompts()
+        module_dict = self._get_prompt_modules()
+        self._validate_prompts(prompts_dict, module_dict)
+
+        # avoid modifying the original dict
+        all_prompts = deepcopy(prompts_dict)
+        for module_name, prompt_module in module_dict.items():
+            # append module name to each key in sub-modules by ":"
+            for key, prompt in prompt_module.get_prompts().items():
+                all_prompts[f"{module_name}:{key}"] = prompt
+        return all_prompts
+
+    def update_prompts(self, prompts_dict: Dict[str, BasePromptTemplate]) -> None:
+        """Update prompts.
+
+        Other prompts will remain in place.
+
+        """
+        prompt_modules = self._get_prompt_modules()
+
+        # update prompts for current module
+        self._update_prompts(prompts_dict)
+
+        # get sub-module keys
+        # mapping from module name to sub-module prompt keys
+        sub_prompt_dicts: Dict[str, PromptDictType] = defaultdict(dict)
+        for key in prompts_dict:
+            if ":" in key:
+                module_name, sub_key = key.split(":")
+                sub_prompt_dicts[module_name][sub_key] = prompts_dict[key]
+
+        # now update prompts for submodules
+        for module_name, sub_prompt_dict in sub_prompt_dicts.items():
+            if module_name not in prompt_modules:
+                raise ValueError(f"Module {module_name} not found.")
+            module = prompt_modules[module_name]
+            module.update_prompts(sub_prompt_dict)
+
+    @abstractmethod
+    def _get_prompts(self) -> PromptDictType:
+        """Get prompts."""
+
+    @abstractmethod
+    def _get_prompt_modules(self) -> PromptMixinType:
+        """Get prompt sub-modules.
+
+        Return a dictionary of sub-modules within the current module
+        that also implement PromptMixin (so that their prompts can also be get/set).
+
+        Can be blank if no sub-modules.
+
+        """
```

Comment:
Should we maybe just have this return empty dict? Majority of the implementation is doing this anyways.

Reply:
weren't you suggesting to leave as abstract method, or was this mostly a comment regarding the prompts

Reply:
oooh ya sorry wasn't clear, last time was mostly about prompts.

Basically looking at the code right now, we mostly override this method at the `Base*` level anyways, so it wouldn't make that much difference (in terms of final component implementation)